### DISCUSSION TITLE
Sprint 6

### DISCUSSION THREAD

**Discussion by Justus-at-Tazama** [tazama-lf/General-Issues#170] [General]
## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

### @cshezi - Cebolenkosi Shezi

**Yesterday:**

**Today:**

### @vorsterk - Kyle Vorster

**Yesterday:**

**Today:**

### @Lenbkr - Len Bekker

**Yesterday:**

**Today:**

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:**

**Today:**

### @rtkay123 - Rodney Kanjala

**Yesterday:**

**Today:**

### @johanfol - Johan Foley

**Yesterday:**

**Today:**

**Comment by Justus-at-Tazama** [comment]
## 2023/07/17

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Eventful Friday - debugging timing attributes. Data Preparation and CRSP were not consistently posting. DP is posting, CRSP is not posting yet. Will be posting the two fields through a "metadata" object in the payload. PR created for Main/Rest branches. Approved and merged.

**Today:** Integrate the metadata into the remaining processors to post at the end. Demo preparation.

### @vorsterk - Kyle Vorster

**Yesterday:** Working on NATS deployment to create the Jenkins jobs. Running jmeter on the cluster - jmeter is outdated on the Cluster, creating issues. New jmeter starter kit acquired and being reviewed. Problem with Redis - we'd previously installed a clustered version, but there is a problem in setting the key. Cluster now removed and single-instance deployed. Debugging end-to-end issue in Redis.

**Today:** Getting the end-to-end up. Demo prep.

### @Lenbkr - Len Bekker

**Yesterday:** Reviewed DataCache and Hashing changes from Data Preparation Service to move everything into the Main/Rest branches. Some bug-fixes in the rules. Reviewed the DataCache issue - Justus was the problem. Recreated the indexes into Arango via Postman. Unit testing via debug-logger.

**Today:** Demo preparation. Add a configurable debug method in the library to force logging when required.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Fixed an issue on the NATS library. When you start up the library the details used will be logged. Reviewed the merge requests with Len to update the main branch. Documentation on the library almost complete.

**Today:** Demo preparation.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Fixed tests for the typology processor to allow merging into main. Reviewed Rule 090. Redeployed, fixed. Reviewed PRs. Assisted with metadata object.

**Today:** Demo prep. Fix end-to-end.

### @johanfol - Johan Foley

**Yesterday:** Trying to get end-to-end up on SB1. Task failed successfully.

**Today:** End-to-end up and running on dev environment.

## Parking lot

Debugging switching. FRMS logger service does not have a logger service, but should not post all the time. Suggested to use log4js. No additional switching flag required.

Tasks for DevOps

@JeanPierreNell - please get end-to-end up on SB1. CRSP appears to be broken. There may also be a caching issue in CADProc.

**Comment by Justus-at-Tazama** [comment]
## 2023/07/17

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Working on propagating the metadata object and timings changes into the MAIN branch. Interfaces embedded in the logic, but needed to be moved to the library instead. Testing in the timings had some unexpected results. Timings appear to have rounding errors, but this doesn't make sense given the calculations. Completed merging to MAIN. Redeployed processors onto SB1. Changes didn't "take". (Foley says there's an issue in deployment of a processor via Jenkins without OpenFaaS where a processor is not automatically restarted - processors must be manually restarted after deployment).

**Today:** Complete the timings across the end-to-end.

### @vorsterk - Kyle Vorster

**Yesterday:** Working on performance benchmarking. Jmeter issues. Scaled up to 3 big-boy nodes (Arango, Elastic, NATS). Spin up as required to save costs. Benchmark tests.

**Today:** Benchmarking all day.

### @Lenbkr - Len Bekker

**Yesterday:** Reviewed debug messages in the library. Added log4js. Need to know if the log level should be drawn from an environment variable. (Add one if there isn't one). Supported Jmeter issues: update of Java from 14 to 21. Updated Jmeter to 5.6.2. Ensured Java dependency compliance. Data Preparation Service shat the bed. Investigating. Request to create two separate images for NATS/REST deployments.

**Today:** Data Preparation Service. Environment switch for log4js loglevel.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Fixed end-to-end. Investigating an error state in NATS - unable to replicate right now on local or see the behaviour on SB1. Please document recreation steps if the error re-occurs. Foley suggests it might be related to a redeployment of NATS. Kyle will test on SB1.

**Today:** Need work.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Working with Shezi on the metadata propagation. Jenkins deployment issue on SB4 (expired token). Investigating unhandled errors in the rule processors. Branch chaos.

**Today:** Need work. (Unhandled errors).

### @johanfol - Johan Foley

**Yesterday:** End-to-end testing. JMeter issues.

**Today:** Benchmarking. Scaling up pods for concurrency and throughput.

## Parking lot

Testing JMeter with NATS. How do you measurement of the throughput. Updated strategy on measuring performance on the platform now that we're using NATS.

REST-based testing of NATS via Postman.

NATS testing, timekeeping.

**Comment by Justus-at-Tazama** [comment]
## 2023/07/19

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Tracing the lost `metaData` object through the flow. Logging the result at various steps to see where it gets lost. Appears to be somewhere between CADProc/TADProc.

**Today:** 

### @vorsterk - Kyle Vorster

**Yesterday:** Started performance runs. The network map is the full typologies. Current figure 130TPS. Timeouts on logstash. For comparison purposes, previous run at 134TPS was 20 big boy servers and only typology 28. We are now only running across 3 big-boy servers. Added the rollout restart kubernetes command on the jenkins jobs for each deployment to restart the pod once deployed.
Added a variable on Jenkins to specify replicas for each processor and rule.

**Today:** Tweaking continues.

### @Lenbkr - Len Bekker

**Yesterday:** Working on getting the end-to-end up and running. Great teamwork. Redeploy every rule to absorb an update to the rule executer. Missing kubectl credentials on SB1. Debugging a rule and seeing some spammy behaviour from NATS.

**Today:** Complete the debug logging via log4js. APM best practices and design.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Worked on the REST-based testing tool for NATS. Waiting for approval on the approach. Checking in with Kyle to get the Jenkins job to deploy the update. Foley will assist. Started implementing APM to work on the end-to-end timestamping via APM.

**Today:** Deploy REST-based testing; carry on with the APM implementation.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Supported the deployment of the end-to-end. Supported Shezi on the metadata object. Reviewed PR for CRSP. Check in on APM to catch up on ELK knowledge.

**Today:** ELK knowledge. Focus on APM and Graphana.

### @johanfol - Johan Foley

**Yesterday:** Supported the debugging of the rule processor results. Found and repaired an issue in the rule executer. NATS meeting - super nice crowd.

**Today:** Benchmarking! Assist on REST-based NATS testing deployment.

## Parking lot

**Comment by Justus-at-Tazama** [comment]
## 2023/07/20

## Attendees
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
@johanfol - Johan Foley

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Processor timings. Testing the processors. Spun up a NATS server on local. Able to run full e2e on local. Picked up a typo on CADProc. PR sent and merged. End-to-end for SB1 is dead. Keeping processor times as raw data and will render in reporting.

**Today:** Complete processor timings.

### @vorsterk - Kyle Vorster

**Yesterday:** Lots of changes on EKS - updated the number of workers on logstash from 5 to 20. No more errors and timeouts. Performance tuning. Added redis cluster. Puzzled over scaling up NATS = less speed.

**Today:** Performance benchmarking.

### @Lenbkr - Len Bekker

**Yesterday:** Added debug logger to the library. APM musical chairs. Reviewed some EKS-APM documentation. Meeting set up to collab with JP and Rodney.

**Today:** Integrating NATS via Prometheus.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Completed and deployed the NATS utilities for REST-based testing. Created a NATS implementation image for GitHub.

**Today:** APM for end-to-end span.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Reviewed PRs. Research on APM.

**Today:** Implementing custom spans on all processors.

### @johanfol - Johan Foley

**Today:** Away

## Parking lot

APM roast - please record.

**Comment by JeanPierreNell** [comment]
## 2023/07/21

## Attendees
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
@johanfol - Johan Foley

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Opened PRs for changes to time calculation. Switched to Nano Seconds. Tested. Redeployed everything. Everything going well,  however end to end is down.

**Today:**  Investigate End to End that is down. Continuing with tickets on the backlog. Will Call JP. 

### @vorsterk - Kyle Vorster

**Yesterday:**  Changes to Jenkins deployments. Add Commands for node affinity. Nats/Arango/Elastic on own node. Worked on getting runs out. Redis Blocker, since redis must be clustered now. Spoke to Len, additional configuration required. Started working on the Prometheus tasks. Foley says no to single instance. Tests paused until len is done. 

**Today:**  Getting Prometheus up for Elastic (Nats Monitoring) Probably helping Jp/Rodney. Will set up on SB1 if works on Dev

### @Lenbkr - Len Bekker

**Yesterday:**  Meeting with the cool guys about the decorators for the library. Reviewed PRs for Shezi. Investigated issue for Justus on Data Prep. Justus raised alarms. Kyle asked to Swap work. Blocker on Redis Cluster went to Len. Prometheus went to Kyle. 

**Today:** Finish Redis Cluster Changes.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Meeting with Len and Rodney for APM Structure. Worked on implementing in the library and setting up local environment. Discovered Fleets this morning, and currently researching. PRs for Shezi, debugging help for Justus.

**Today:** Getting APM up and running with fleets. 

### @rtkay123 - Rodney Kanjala

**Yesterday:**  Meeting with Len & JP on APM Structure. PR from Shezi. Investigated why a certain decorator wasn’t working. Paused in favor of setting up local env. Set up successful. JP brought up fleets – Doing a binary search on the version of kibana to find a working version. 

**Today:** Finish setting up the Microservice. Getting the pieces to connect and work on spans.

### @johanfol - Johan Foley

**Today:** Away

## Parking lot

**Comment by Justus-at-Tazama** [comment]
## 2023/07/24

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** SB1 end-to-end. Testing with the Moving Van. Successful. But not working now. Redeploy SB2 with the latest code. Supported by Kyle.

**Today:** Update configs on SB2. Use the Postman collection. Fix SB1 please.

### @vorsterk - Kyle Vorster

**Yesterday:** Supporting the deployment of the platform to SB2 for PoC. Testing today. Pasued performance while changes to Clustered Redis being made. Prometheus for NATS installed. Getting an elastic agent up for the reporting.

**Today:** Resume benchmarking once Clustered Redis is complete.

### @Lenbkr - Len Bekker

**Yesterday:** Updates to the library. Formatting tool broke auto-complete. Environment set up to test against both clustered and unclustered redis for testing. Node redirects conflicting with our design. Failure rate unacceptable. Node may be preferred.

**Today:** Redis.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** SB1 end to end was briefly up and running. Currently not up. Carried on with APM implementation. Fighting with Docker.

**Today:** End-to-end please. Then APM.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Setting up a local environment for ELK. Took a while to secure an enrolment token to connect Elastic and Kibana. Decorators do not appear to be working for statements. Trying a solution today.

**Today:** Consolidate knowledge and implement.

### @johanfol - Johan Foley

**Yesterday:**

**Today:**

## Parking lot

Timebox for end-to-end fix - 90 minutes. (JP) Starting with redeployment of the Data Preparation Service

**Comment by vorsterk** [comment]
## 2023/07/25

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** SB2 for e2e , network map and config on arango

**Today:** Run bankserv data , and e2e for SB2

### @vorsterk - Kyle Vorster

**Yesterday:** Upgraded the Kubernetes version on SB1 and SB2 , due to them being outdated , Had to change a few volumes after the upgrade , arangodb was outdated after the upgrade , updated arango on both. Helped Len with redis cluster . Added the envname to the docker image in Sb1

**Today:** Carry on with Sb3 and 4 upgrade , add nats (includes Jenkins and the deployments) and check that E2e works . Once Len is finished with the Redis changes I will run a performance test.

### @Lenbkr - Len Bekker

**Yesterday:** Redis cluster working , Kyle helped , exposing the endpoints for e2e service . Got it to connect to single and cluster in the code. 

**Today:** Port Redis to Lib today , to use it across everything . 

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Looking at e2e on sb1 , investigating DP , APM 

**Today:** Check e2e today , checking Justus request , APM – breakdown , wait time ie: reporting’s 

### @rtkay123 - Rodney Kanjala

**Yesterday:** Working on adding custom spans to the processors , decorators added , flag needs to be raise config issue , 

**Today:** Continue on custom spans for APM , need more info about the extensive reporting , Foley gave the info (span for anything external)

### @johanfol - Johan Foley

**Yesterday:** Looking at DP trying to get E2E working on SB1

**Today:** Look at the bankserv data through sandonx2 , Sybrin reports – update on Sybrin side , draw some metrics for that .

## Parking lot

JP – sb1 – how we catch errors, DP throwing errors when data cache isn’t there , when it restarts .  Index.ts – checking for incoming and outcome. Fix bug in DP

**Comment by JeanPierreNell** [comment]
## 2023/07/26

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Worked on SB2 to get e2e up and running. Success! Moved on to batch-ppa, had to remove openfaas. Asked Rodney for assistance. Kyle helped with files to include on batch-ppa. Issues with docker security, batch-ppa is accessing it's own file, had to give premission. Foley and Len assisted with permissions. Success! Foley was working on CMS, which will report to Sybrin. 

**Today:** Will continue sending batches. If can complete, will move on to the next. 

### @vorsterk - Kyle Vorster

**Yesterday:** Finished upgrading SB 3 . Removed Arango CLuster, removed Ni-fi, added NATS with configuration. Added Jenkins config for nats - everything deployed, need to test e2e. JP mentioned CRSP broken. Everything connected to NATS on SB3. SB4 had the same treatment. Had to update jenkins on SB 3 + 4. 

**Today:** Deploying Rules and processors on SB4. Rodney wants to use it. If Len finish redis, can run performance tests.

### @Lenbkr - Len Bekker

**Yesterday:** Made changes on all processors using Redis to incorporate redis clusters. Len and JP will investigate CRSP issues. Will test cluster mode once e2e is up. Looked into way to monitor NATS, where we can view the messages

**Today:** Still busy on SB1 with Redis. 

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Worked on broken NATS promise once the processor errors. Struggled with the eslint. approached it different ways - chose the most elegant solution (I think). Will open to the floor. 

**Today:** Fix CRSP, upgrade all processors

### @rtkay123 - Rodney Kanjala

**Yesterday:** Adding Custom Spans to rule processors. Started with Data-Prep. Len was breaking stuff on SB1, SB 3 + 4 was being upgraded. Implemented APM on a few processors - untested. Needed to make changes to data-prep to compensate for a drop in coverage. Jest didn't complain when trying to push with to little coverage in unit tests. Helped Shezi with issues on batch-ppa.

**Today:** Test all APM canges. Implemented APM on the top level. Need a discussion on how to propagate the parent spans. Linking spans across microservices.

### @johanfol - Johan Foley

**Yesterday:** Absent

**Today:** Absent

## Parking lot

Timebox for end-to-end fix - 90 minutes. (JP) Starting with redeployment of the Data Preparation Service

**Comment by JeanPierreNell** [comment]
## 2023/07/27

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Worked on SB2 to get e2e up and running. Success! Moved on to batch-ppa, had to remove openfaas. Asked Rodney for assistance. Kyle helped with files to include on batch-ppa. Issues with docker security, batch-ppa is accessing it's own file, had to give premission. Foley and Len assisted with permissions. Success! Foley was working on CMS, which will report to Sybrin. 

**Today:** Will continue sending batches. If can complete, will move on to the next. 

### @vorsterk - Kyle Vorster

**Yesterday:** Worked with JP and Len – Clustered Redis is up and running. Experts said don’t write to disk on Redis. Updated on Dev. APM spins up , but loses context of the elastic cert. 

**Today:** Spinning up 3 big boys and Running a test now. Checking why redis connections gets dropped. Need to deploy different changes to SBs

### @Lenbkr - Len Bekker

**Yesterday:** Ensured Redis-Cluster changes working E2E on SB1 for both redis with cluster mode enabled or not. Fixed broken tests and bugs with JP. Deployed third party nats web testing util.

**Today:** Move deployments on SB1 back to main branch after merge. Help Kyle where possible so he can get everything on DEV env working with latest changes to run benchmark.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Worked on Getting Sandbox 1 working end-to-end. Fixed broken tests, worked on bugs in the process. Upgraded library Versions.

**Today:** Assist Kyle with APM

### @rtkay123 - Rodney Kanjala

**Yesterday:** Working on APM setup. Waiting for Kyles deployments. Followed up on why husky didn’t care for test coverage.

**Today:** Figuring out why husky doesn’t care about test coverage. 

### @johanfol - Johan Foley

**Yesterday:** Worked on Getting SB1 through.

**Today:** Helping the team out. 

## Parking lot
